### PR TITLE
Add tosa -> rock.attention conversion

### PR DESCRIPTION
**Pull request by manupak**
This commit adds frontend op rock.attention
and creates conversions from tosa.

closes : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/959

### CODE REVIEW

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 1]
```diff
@@ -202,6 +202,43 @@ def Rock_ReduceOp :
   }];
 }
 
+def Rock_AttentionOp :
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td:205]: Can we implement `RockGemmWrapperInterface` for this, even if in a somewhat artificial way?

**Review by krzysz00 — COMMENTED** [review]
I don't have any issues with the PR itself, but I would like to understand, for example, how we'll be tuning a `rock.attention`

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 1 above)
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:205]: RockGemmWrapperInterface seems specific to gemm-based kernels (and their interface methods).
Hence, I dont want to bring that here now. For the bring-up, we can set a universal param whilst rock.attention op could accept a "perf-config". I'd like to sort the tuning after we have something functional.

Going forward, one idea might be to introduce a RockTunableOpInterface that could prescribe the following:
1) op->getTuningRange(struct TunableParams* tuningSpace)
2) op->tuningSetParam(ParamEntry* paramEntry)

Comment (krzysz00) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:205]: Ok, the only methods specific to gemms on `RockGemmWrapperInterface` are `getAType` and `getBType`. If it weren't for those, we could actually straight-up go ahead and rename the darn thing to `RockTunableKernelInterface` or something.

Given the constraints on flash attention, I'd be happy with an implementation that sets `getAType() == getBType() == [get the type of something in there]`.

On the third hand, I think there's an argument that we could have, for example `queries : fp8, keys : bf8, values : f32` or `queries : int8, keys : int8, values : int32` or something like that, so perhaps `getAType() = getElementType(queries)` and `getBType = getElementType(keys)`, but I don't fully know the usage context.

So I think there's an argument, for now, for using the inner GEMM (queries * keys) as the gemm for RockGemmWrapperInterface, which would let us plug `attention` into the high-level flows along with `conv` and `gemm`.

Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:205]: Lets move this discussion to : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/954.
I dont think this is relevant to this PR.

Comment (krzysz00) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:205]: Agreed, and agreed that there's a longer design discussion to be had here that is out of scope for this PR.

**Review by krzysz00 — COMMENTED** [review]
Approved per earlier comments, deferral of tuning discussion to a later date

**Review by krzysz00 — APPROVED** [review]
... this time with actually hitting the approve setting lol

**Pull request merged**